File: app/src/main/java/com/veganbeauty/app/features/account/notification/NotificationListAdapter.java
package com.veganbeauty.app.features.account.notification;

import android.animation.ObjectAnimator;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.ColorStateList;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.veganbeauty.app.R;
import com.veganbeauty.app.data.local.entities.NotificationItem;
import com.veganbeauty.app.databinding.AccountNotificationHeaderBinding;
import com.veganbeauty.app.databinding.AccountNotificationItemBinding;

public class NotificationListAdapter extends ListAdapter<NotificationListItem, RecyclerView.ViewHolder> {

    private static final int VIEW_TYPE_HEADER = 0;
    private static final int VIEW_TYPE_ITEM = 1;

    public interface OnNotificationClickListener {
        void onClick(NotificationItem item);
    }

    private static final int MAX_CONTENT_WORDS = 200;

    private final OnNotificationClickListener onItemClick;
    private final OnNotificationClickListener onMarkReadClick;
    private final OnNotificationClickListener onDeleteClick;

    // Track currently open swipe item
    private NotificationViewHolder currentOpenHolder = null;

    public NotificationListAdapter(
            OnNotificationClickListener onItemClick,
            OnNotificationClickListener onMarkReadClick,
            OnNotificationClickListener onDeleteClick
    ) {
        super(new DiffCallback());
        this.onItemClick = onItemClick;
        this.onMarkReadClick = onMarkReadClick;
        this.onDeleteClick = onDeleteClick;
    }

    private static String truncateWords(String text, int maxWords) {
        if (text == null) return "";
        String[] words = text.trim().split("\\s+");
        if (words.length <= maxWords) return text;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < maxWords; i++) {
            if (i > 0) sb.append(' ');
            sb.append(words[i]);
        }
        sb.append("...");
        return sb.toString();
    }

    /** Snap the currently-open item back to closed position */
    public void closeOpenedItem() {
        if (currentOpenHolder != null) {
            currentOpenHolder.animateForeground(0f);
            currentOpenHolder = null;
        }
    }

    /** Called by the Fragment's touch listener to open/close the swiped item */
    public void openItem(NotificationViewHolder holder, float actionWidthPx) {
        if (currentOpenHolder != null && currentOpenHolder != holder) {
            currentOpenHolder.animateForeground(0f);
        }
        holder.animateForeground(-actionWidthPx);
        currentOpenHolder = holder;
    }

    @Override
    public int getItemViewType(int position) {
        NotificationListItem item = getItem(position);
        if (item instanceof NotificationListItem.Header) {
            return VIEW_TYPE_HEADER;
        } else {
            return VIEW_TYPE_ITEM;
        }
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        if (viewType == VIEW_TYPE_HEADER) {
            AccountNotificationHeaderBinding binding = AccountNotificationHeaderBinding.inflate(inflater, parent, false);
            return new HeaderViewHolder(binding);
        } else {
            AccountNotificationItemBinding binding = AccountNotificationItemBinding.inflate(inflater, parent, false);
            return new NotificationViewHolder(binding);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        NotificationListItem item = getItem(position);
        if (item instanceof NotificationListItem.Header) {
            ((HeaderViewHolder) holder).bind((NotificationListItem.Header) item);
        } else if (item instanceof NotificationListItem.Notification) {
            ((NotificationViewHolder) holder).bind(((NotificationListItem.Notification) item).getItem());
        }
    }

    public static class HeaderViewHolder extends RecyclerView.ViewHolder {
        private final AccountNotificationHeaderBinding binding;

        public HeaderViewHolder(AccountNotificationHeaderBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(NotificationListItem.Header header) {
            binding.tvHeaderTitle.setText(header.getTitle());
        }
    }

    public class NotificationViewHolder extends RecyclerView.ViewHolder {
        private final AccountNotificationItemBinding binding;

        public NotificationViewHolder(AccountNotificationItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public View getForeground() { return binding.cardNotification; }

        /** Animate the foreground card to a target X translation */
        public void animateForeground(float targetX) {
            ObjectAnimator anim = ObjectAnimator.ofFloat(binding.cardNotification, "translationX", binding.cardNotification.getTranslationX(), targetX);
            anim.setDuration(180);
            anim.start();
        }

        public void bind(NotificationItem item) {
            Context context = binding.getRoot().getContext();

            // Reset swipe state on re-bind
            binding.cardNotification.setTranslationX(0f);

            binding.tvTitle.setText(item.getTitle());
            binding.tvContent.setText(truncateWords(item.getContent(), MAX_CONTENT_WORDS));
            binding.tvTime.setText(NotificationDateHelper.getDisplayTime(item.getTime()));

            binding.viewUnreadDot.setVisibility(item.isRead() ? View.GONE : View.VISIBLE);
            binding.cardNotification.setCardBackgroundColor(
                    item.isRead()
                            ? ContextCompat.getColor(context, R.color.white)
                            : android.graphics.Color.parseColor("#E8F5E9")
            );

            int resId = resolveNotificationIconRes(context, item.getIconResName());
            if (resId != 0) {
                binding.ivIcon.setImageResource(resId);
            } else {
                binding.ivIcon.setImageResource(R.drawable.ic_voucher);
            }

            int iconTint;
            int bgTint;

            if ("Khuyến mãi".equals(item.getCategory())) {
                iconTint = ContextCompat.getColor(context, R.color.status_pending_text);
                bgTint = ContextCompat.getColor(context, R.color.status_pending_bg);
            } else if ("Đơn hàng".equals(item.getCategory())) {
                iconTint = ContextCompat.getColor(context, R.color.status_delivering_text);
                bgTint = ContextCompat.getColor(context, R.color.status_delivering_bg);
            } else if ("Khác".equals(item.getCategory())) {
                iconTint = ContextCompat.getColor(context, R.color.status_processing_text);
                bgTint = ContextCompat.getColor(context, R.color.status_processing_bg);
            } else {
                iconTint = ContextCompat.getColor(context, R.color.primary);
                bgTint = ContextCompat.getColor(context, R.color.gray_light);
            }

            binding.ivIcon.setColorFilter(iconTint);
            binding.ivIcon.setBackgroundTintList(ColorStateList.valueOf(bgTint));

            // Foreground tap → open notification
            binding.cardNotification.setOnClickListener(v -> {
                if (binding.cardNotification.getTranslationX() != 0) {
                    closeOpenedItem();
                    return;
                }
                if (onItemClick != null) onItemClick.onClick(item);
            });

            // Action button clicks (revealed on swipe left)
            binding.btnMarkReadAction.setOnClickListener(v -> {
                closeOpenedItem();
                if (onMarkReadClick != null) onMarkReadClick.onClick(item);
            });
            binding.btnDeleteAction.setOnClickListener(v -> {
                closeOpenedItem();
                if (onDeleteClick != null) onDeleteClick.onClick(item);
            });
        }
    }

    private static int resolveNotificationIconRes(Context context, String iconResName) {
        if (iconResName == null || iconResName.isEmpty()
                || "ic_bell".equals(iconResName)
                || "home_ic_bell".equals(iconResName)) {
            return R.drawable.ic_bell;
        }
        int resId = context.getResources().getIdentifier(iconResName, "drawable", context.getPackageName());
        return resId != 0 ? resId : R.drawable.ic_bell;
    }

    private static class DiffCallback extends DiffUtil.ItemCallback<NotificationListItem> {
        @Override
        public boolean areItemsTheSame(@NonNull NotificationListItem oldItem, @NonNull NotificationListItem newItem) {
            if (oldItem instanceof NotificationListItem.Header && newItem instanceof NotificationListItem.Header) {
                return ((NotificationListItem.Header) oldItem).getTitle().equals(((NotificationListItem.Header) newItem).getTitle());
            } else if (oldItem instanceof NotificationListItem.Notification && newItem instanceof NotificationListItem.Notification) {
                return ((NotificationListItem.Notification) oldItem).getItem().getId().equals(((NotificationListItem.Notification) newItem).getItem().getId());
            }
            return false;
        }

        @SuppressLint("DiffUtilEquals")
        @Override
        public boolean areContentsTheSame(@NonNull NotificationListItem oldItem, @NonNull NotificationListItem newItem) {
            return oldItem.equals(newItem);
        }
    }
}

File: app/src/main/java/com/veganbeauty/app/features/account/notification/NotificationListAdapter.kt
package com.veganbeauty.app.features.account.notification

import android.content.res.ColorStateList
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.veganbeauty.app.R
import com.veganbeauty.app.data.local.entities.NotificationItem
import com.veganbeauty.app.databinding.AccountNotificationHeaderBinding
import com.veganbeauty.app.databinding.AccountNotificationItemBinding

private const val VIEW_TYPE_HEADER = 0
private const val VIEW_TYPE_ITEM = 1

class NotificationListAdapter(
    private val onItemClick: (NotificationItem) -> Unit,
    private val onActionClick: (NotificationItem) -> Unit,
    private val onMarkReadClick: (NotificationItem) -> Unit,
    private val onDeleteClick: (NotificationItem) -> Unit
) : ListAdapter<NotificationListItem, RecyclerView.ViewHolder>(DiffCallback) {

    override fun getItemViewType(position: Int): Int {
        return when (getItem(position)) {
            is NotificationListItem.Header -> VIEW_TYPE_HEADER
            is NotificationListItem.Notification -> VIEW_TYPE_ITEM
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return when (viewType) {
            VIEW_TYPE_HEADER -> {
                val binding = AccountNotificationHeaderBinding.inflate(inflater, parent, false)
                HeaderViewHolder(binding)
            }
            VIEW_TYPE_ITEM -> {
                val binding = AccountNotificationItemBinding.inflate(inflater, parent, false)
                NotificationViewHolder(binding)
            }
            else -> throw IllegalArgumentException("Unknown view type: $viewType")
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (val item = getItem(position)) {
            is NotificationListItem.Header -> (holder as HeaderViewHolder).bind(item)
            is NotificationListItem.Notification -> (holder as NotificationViewHolder).bind(item.item)
        }
    }

    inner class HeaderViewHolder(
        private val binding: AccountNotificationHeaderBinding
    ) : RecyclerView.ViewHolder(binding.root) {
        fun bind(header: NotificationListItem.Header) {
            binding.tvHeaderTitle.text = header.title
        }
    }

    inner class NotificationViewHolder(
        private val binding: AccountNotificationItemBinding
    ) : RecyclerView.ViewHolder(binding.root) {
        
        fun bind(item: NotificationItem) {
            val context = binding.root.context
            
            // Set basic texts
            binding.tvTitle.text = item.title
            binding.tvContent.text = item.content
            binding.tvTime.text = item.time

            // Unread dot visibility
            binding.viewUnreadDot.visibility = if (item.isRead) View.GONE else View.VISIBLE

            // Dynamically load category icon
            val resId = context.resources.getIdentifier(item.iconResName, "drawable", context.packageName)
            if (resId != 0) {
                binding.ivIcon.setImageResource(resId)
            } else {
                binding.ivIcon.setImageResource(R.drawable.ic_voucher)
            }

            // Beautiful dynamic category coloring for premium UI feel
            val (iconTint, bgTint) = when (item.category) {
                "Khuyến mãi" -> Pair(
                    ContextCompat.getColor(context, R.color.status_pending_text),
                    ContextCompat.getColor(context, R.color.status_pending_bg)
                )
                "Đơn hàng" -> Pair(
                    ContextCompat.getColor(context, R.color.status_delivering_text),
                    ContextCompat.getColor(context, R.color.status_delivering_bg)
                )
                "Khác" -> Pair(
                    ContextCompat.getColor(context, R.color.status_processing_text),
                    ContextCompat.getColor(context, R.color.status_processing_bg)
                )
                else -> Pair(
                    ContextCompat.getColor(context, R.color.primary),
                    ContextCompat.getColor(context, R.color.gray_light)
                )
            }
            binding.ivIcon.setColorFilter(iconTint)
            binding.ivIcon.backgroundTintList = ColorStateList.valueOf(bgTint)

            // Dynamic Action Button visibility & styling
            if (!item.actionText.isNullOrEmpty() && !item.actionText.equals("CHI TIẾT", ignoreCase = true) && !item.actionText.equals("CHI TIẾT ĐƠN", ignoreCase = true)) {
                binding.btnAction.visibility = View.VISIBLE
                binding.btnAction.text = item.actionText
                binding.btnAction.setOnClickListener {
                    onActionClick(item)
                }
            } else {
                binding.btnAction.visibility = View.GONE
            }

            // Bind read/delete clicks
            binding.btnMarkRead.visibility = if (item.isRead) View.GONE else View.VISIBLE
            binding.btnMarkRead.setOnClickListener {
                onMarkReadClick(item)
            }

            binding.btnDelete.setOnClickListener {
                onDeleteClick(item)
            }

            // Card click listener
            binding.cardNotification.setOnClickListener {
                onItemClick(item)
            }
        }
    }

    private companion object DiffCallback : DiffUtil.ItemCallback<NotificationListItem>() {
        override fun areItemsTheSame(oldItem: NotificationListItem, newItem: NotificationListItem): Boolean {
            return when {
                oldItem is NotificationListItem.Header && newItem is NotificationListItem.Header -> 
                    oldItem.title == newItem.title
                oldItem is NotificationListItem.Notification && newItem is NotificationListItem.Notification -> 
                    oldItem.item.id == newItem.item.id
                else -> false
            }
        }

        override fun areContentsTheSame(oldItem: NotificationListItem, newItem: NotificationListItem): Boolean {
            return oldItem == newItem
        }
    }
}
